### PR TITLE
fix+update transducer

### PR DESCRIPTION
**Pull request by b-flo**
Hi,

Sorry for the delay. This PR update a bunch of things :

- [x] add backward compatibility (disabled after unification)
- [x] add missing bias initialization for RNN parts
- [x] add model loss averaging (small hack)
- [x] fix transfer learning (related to #1654) 
- [x] fix voxforge run script + add transformer support (related to #1654)
- [x] fix vivos results + model update (related to #1689)
- [x] fix default transfer learning arguments (related to #1654)
- [x] add documentation and tutorial for transducers. (related to #1682)
          -> Edit: @sw005320 or @kan-bayashi could you take a look at this part please? I think I'll keep it minimal for now and extend with report for all architectures in next PR.

Edit: Not included:
X add librispeech recipe for T-T and T/RNN-T (related to #1704)
X add voxforge recipe for T-T and T/RNN-T
X add transformer LM pre-init for transformer decoder

@hirofumi0810 Could you take a look at the change in `asr_init` please? I reverted to the state before your changes in [eb5ab0f](https://github.com/espnet/espnet/commit/eb5ab0f095fa906573485e5df9d569b82b3e148a#diff-a4474b45220fd32f6849a023ceffac63) but I'm not sure what was the intention? Transformer pre-init works with the old version in my case and for the new version logs would not be displayed to user if a key is missing in state_dict (because of `model_state_dict[key_p].shape`)

### CODE REVIEW

**Comment by hirofumi0810** [comment]
> Hi,
> 
> Sorry for the delay. This PR update a bunch of things :
> 
>     * [x]  add backward compatibility (disabled after unification)
> 
>     * [x]  add missing bias initialization for RNN parts
> 
>     * [x]  add model loss averaging (small hack)
> 
>     * [x]  fix transfer learning (related to #1654)
> 
>     * [x]  fix voxforge run script + add transformer support (related to #1654)
> 
>     * [x]  fix vivos results + model update (related to #1689)
> 
>     * [x]  fix default transfer learning arguments (related to #1654)
> 
>     * [ ]  add documentation and tutorial for transducers. (related to #1682)
> 
>     * [ ]  add librispeech recipe for T-T and T/RNN-T (related to #1704)
> 
>     * [ ]  add voxforge recipe for T-T and T/RNN-T
> 
>     * [ ]  add transformer LM pre-init for transformer decoder
> 
> 
> @hirofumi0810 Could you take a look at the change in `asr_init` please? I reverted to the state before your changes in [eb5ab0f](https://github.com/espnet/espnet/commit/eb5ab0f095fa906573485e5df9d569b82b3e148a#diff-a4474b45220fd32f6849a023ceffac63) but I'm not sure what was the intention? Transformer pre-init works with the old version in my case and for the new version logs would not be displayed to user if a key is missing in state_dict (because of `model_state_dict[key_p].shape`)

@b-flo 
As far as I remember, there was no bug to start the training in the oldest version, but maybe that did not initialized the model correctly.
I think I confirmed it by looking at the training log in the early training stage.

**Comment by b-flo** [comment]
> @b-flo
> As far as I remember, there was no bug to start the training in the oldest version, but maybe that did not initialized the model correctly.
> I think I confirmed it by looking at the training log in the early training stage.

If there is any mismatch, keys for both state dict will be printed with a warning message, and model will not be updated. However, I assumed the training should continue even if the pre-training part is discarded. That may be the reason?

**Comment by b-flo** [comment]
Hi,

I think this PR is done. I'll add the remaining items in another PR, alongside the new transformer-based models I talked about on slack.

Also, I guess I should remove scripts related to transducer and transfer learning in voxforge. It is honestly to much work to maintain while the showcased options are easy to use in default setting.

**Comment by kamo-naoyuki** [comment]
@b-flo Would you merge master to fix ci?

**Comment by kamo-naoyuki** [comment]
@sw005320 Is this PR is good? I'll apply black after this PR. Now seems to be good for it.

**Comment by sw005320** [comment]
Can you wait for one day?

**Comment by b-flo** [comment]
I agree on that but I'll most likely remove all voxforge scripts in the next PR for the following reasons:

- The scripts were created to demonstrate transducer and transfer learning usage however most of the complexity come from the fact I handle pre-training steps + on-the-fly config creation. This is so much complex compare to how easy it is to use transducer or/and transfer learning "normally".
- Maintaining these scripts will be really hard given I'll add 1) new encoder and decoder architecture and 2) full customization on transformer architecture definition (e.g: definition of each layers in each blocks instead of repeated blocks based on single arguments)

I was thinking of keeping these scripts for RNN-based architectures only but I have work in progress related to RNN that will require changing them again...

**Comment by sw005320** [comment]
OK.
This is no problem.
Many thank @b-flo!

**Comment by ahmedalbahnasawy** [comment]
@b-flo I changed Librispeech recipe to run RNN-T model. I uploaded the trained model to the drive.

**Comment by b-flo** [comment]
Hi @ahmedalbahnasawy,

Thank you, I'll take a look! Also, could you share your results please?
